File: 2015/python/d21.py
weapons = [[8, 4, 0], [10, 5, 0], [25, 6, 0], [40, 7, 0], [74, 8, 0]]
armour = [[0, 0, 0], [13, 0, 1], [31, 0, 2], [53, 0, 3], [75, 0, 4], [102, 0, 5]]
rings = [[0, 0, 0], [25, 1, 0], [50, 2, 0], [100, 3, 0], [20, 0, 1], [40, 0, 2], [80, 0, 3]]

def read(filename):
    for line in open(filename, 'r'):
        line = line.replace('\n', '').split(": ")
        val = int(line[1])
        if line[0] == 'Hit Points':
            hp = val
        elif line[0] == 'Damage':
            dam = val
        else:
            arm = val
    return (hp, dam, arm)

def fight(hp, dam, arm, boss_hp, boss_dam, boss_arm):
    while True:
        boss_hp -= max(1, dam - boss_arm)
        if (boss_hp <= 0):
            return True
        hp -= max(1, boss_dam - arm)
        if (hp <= 0):
            return False

def part1(boss_hp, boss_dam, boss_arm):
    global weapons, armour, rings
    best = 9999
    for w in weapons:
       for a in armour:
          for r1 in rings:
              for r2 in rings:
                  if (r1[0] == r2[0]) and (r1[0] != 0):
                      continue
                  cost = w[0] + a[0] + r1[0] + r2[0]
                  if cost >= best:
                      continue
                  dam = w[1] + r1[1] + r2[1]
                  arm = a[2] + r1[2] + r2[2]
                  if (fight(100, dam, arm, boss_hp, boss_dam, boss_arm)):
                      best = cost
    return best

def part2(boss_hp, boss_dam, boss_arm):
    global weapons, armour, rings
    best = 0
    for w in weapons:
       for a in armour:
          for r1 in rings:
              for r2 in rings:
                  if (r1[0] == r2[0]) and (r1[0] != 0):
                      continue
                  cost = w[0] + a[0] + r1[0] + r2[0]
                  if cost < best:
                      continue
                  dam = w[1] + r1[1] + r2[1]
                  arm = a[2] + r1[2] + r2[2]
                  if (not fight(100, dam, arm, boss_hp, boss_dam, boss_arm)):
                      best = cost
    return best

def tests():
  res = fight(8, 5, 5, 12, 7, 2)
  assert(res == 1)

tests()
(boss_hp, boss_dam, boss_arm) = read("../inputs/input_21.txt")
print(part1(boss_hp, boss_dam, boss_arm))
print(part2(boss_hp, boss_dam, boss_arm))

File: 2015/R/d21.R
shop <- data.frame(
  cost = c(8,10,25,40,74,13,31,53,75,102,0,25,50,100,20,40,80,0,0),
  dmg = c(4,5,6,7,8,0,0,0,0,0,0,1,2,3,0,0,0,0,0),
  arm = c(0,0,0,0,0,1,2,3,4,5,0,0,0,0,1,2,3,0,0),
  type = c(1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,3,3))
shop <- split(shop, shop$type)

d <- read.csv(text = readLines("../inputs/input_21.txt"), sep = ":", 
              header = FALSE, col.names = c("n", "v"))
enemy <- data.frame(
  hp = d$v[d$n == "Hit Points"],
  dmg = d$v[d$n == "Damage"],
  arm = d$v[d$n == "Armor"])

solve <- function() {
  nwea <- nrow(shop[[1]])
  narm <- nrow(shop[[2]])
  nrng <- nrow(shop[[3]])
  tries <- data.frame(
    wea = rep(seq_len(nwea), each = narm * nrng * nrng),
    arm = rep(seq_len(narm), each = nrng * nrng),
    r1 = rep(seq_len(nrng), each = nrng),
    r2 = seq_len(nrng)
  )
  tries <- tries[tries$r2 > tries$r1,]
  tries$cost <- shop[[1]]$cost[tries$wea] + shop[[2]]$cost[tries$arm] + 
                shop[[3]]$cost[tries$r1] + shop[[3]]$cost[tries$r2]
  tries$tdmg <- shop[[1]]$dmg[tries$wea] +  
              shop[[3]]$dmg[tries$r1] + shop[[3]]$dmg[tries$r2]
  tries$tarm <- shop[[2]]$arm[tries$arm] +  
                shop[[3]]$arm[tries$r1] + shop[[3]]$arm[tries$r2]
  tries$win <- pmax(1, tries$tdmg - enemy$arm) >= pmax(1, enemy$dmg - tries$tarm)
  c(min(tries$cost[tries$win]), max(tries$cost[!tries$win]))
}  
  
solve()
